test/externalTests/external_test.py

```diff
@@ -0,0 +1,364 @@
+    solc: Dict[str, SolcConfig] = field(
+        default_factory=lambda: defaultdict(SolcConfig))
+
+    def __post_init__(self):
+        if isinstance(self.solc, dict):
+            self.solc = SolcConfig(**self.solc)
+
+    def selected_presets(self):
+        return self.compile_only_presets + self.settings_presets
+
+
+class TestRunner(metaclass=ABCMeta):
+
+    @staticmethod
+    def on_local_test_dir(fn):
+        """Run function inside the test directory"""
+
+        def f(self, *args, **kwargs):
+            if self.test_dir:
+                os.chdir(self.test_dir)
+            else:
+                raise RuntimeError("Test directory not defined")
+            return fn(self, *args, **kwargs)
+        return f
+
+    @abstractmethod
+    def setup_environment(self, test_dir: Path):
+        pass
+
+    @abstractmethod
+    def clean(self):
+        pass
+
+    @abstractmethod
+    def compiler_settings(self, solc_version: str, presets: Tuple[str] = AVAILABLE_PRESETS):
+        pass
+
+    @abstractmethod
+    def compile(self, solc_version: str, preset: str):
+        pass
+
+    @abstractmethod
+    def run_test(self, preset: str):
+        pass
+
+
+class ExternalTest:
+    def __init__(self, config: TestConfig):
+        if config.solc.binary_type not in ('native', 'solcjs'):
+            raise RuntimeError(
+                f"Invalid solidity compiler binary type: {config.solc.binary_type}")
+        if config.solc.binary_type != "solcjs" and config.solc.solcjs_src_dir != "":
+            raise RuntimeError(
+                f"""Invalid test configuration: 'native' mode cannot be used with 'solcjs_src_dir'.
+              Please use 'binary_type: solcjs' or unset: 'solcjs_src_dir: {config.solc.solcjs_src_dir}'""")
+        self.config = config
+
+    @staticmethod
+    def parse_command_line(description: str, args: str):
+        arg_parser = ArgumentParser(description)
+        arg_parser.add_argument('--solc-binary-type', required=True, type=str,
+                                help="""Solidity compiler binary type""",
+                                choices=['native', 'solcjs'])
+        arg_parser.add_argument('--solc-binary-path', required=True, type=str,
+                                help="""Path to solc or soljson.js binary""")
+        return arg_parser.parse_args(args)
+
+    @staticmethod
+    def download_project(test_dir: Path, repo_url: str,
+                         ref_type: str = "branch", ref: str = "master"):
+        if ref_type not in ('commit', 'branch', 'tag'):
+            raise RuntimeError(f"Invalid reference type: {ref_type}")
+
+        print(f"Cloning {ref_type} {ref} of {repo_url}...")
+        if ref_type == "commit":
+            os.mkdir(test_dir)
+            os.chdir(test_dir)
+            run_cmd("git init")
+            run_cmd(f"git remote add origin {repo_url}")
+            run_cmd(f"git fetch --depth 1 origin {ref}")
+            run_cmd("git reset --hard FETCH_HEAD")
```

Comment:
Maybe it would be better to extend [`git_helpers.py`](https://github.com/ethereum/solidity/blob/develop/scripts/common/git_helpers.py) instead?

Reply:
Oh right. I missed that file. I will extend and reuse it.